src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,548 @@
+                (i) => Math.abs(this.getActiveIndex() - i)),
+            true);
+      } else {
+        this.goTo(index, true);
+      }
+    }
+  }
+
+  /**
+   * Returns an integer number for half of the slides. The parameter
+   * specifies whether, if given an array containing an odd number of elements,
+   * the larger odd value for half should be returned, or the smaller even
+   * number.
+   * @param weightOdd
+   * @private
+   */
+  private getHalfOfSlideCount(weightOdd: boolean): number {
+    const halfLength = (this.getSlides().length - 1) / 2;
+    if (halfLength % 2 === 0) {
+      return halfLength;
+    } else if (weightOdd) {
+      return Math.ceil(halfLength);
+    } else {
+      return Math.floor(halfLength);
+    }
+  }
+
+  /**
+   * Return half of the slides by taking slides to one side of the given slide.
+   *
+   * This function assumes looping is enabled and loops around if needed to get
+   * half of the slides.
+   */
+  private splitSlidesInHalf(
+      slide: HTMLElement, direction: Direction
+  ): HTMLElement[] {
+    const targetLength =
+        this.getHalfOfSlideCount(direction === Direction.RIGHT);
+    const result = [];
+    let indexToAdd = this.getIndex(slide);
+    while (result.length < targetLength) {
+      // Build the looped index
+      indexToAdd = mathf.wrap(indexToAdd + direction, 0, this.getSlideCount());
+      const slideToAdd = this.getSlideByIndex(indexToAdd);
+      if (direction === Direction.RIGHT) {
+        result.push(slideToAdd);
+      } else {
+        result.unshift(slideToAdd);
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Setup initial values.
+   */
+  private init(): void {
+    this.transition.init(this);
+    this.raf.start();
+    this.resetAutoplayTimeout();
+  }
+
+  /**
+   * Clear the current autoplay timeout and prep a new one.
+   * @private
+   */
+  private resetAutoplayTimeout() {
+    if (this.autoplaySpeed !== null) {
+      clearTimeout(this.autoplayTimeout);
+      this.autoplayTimeout =
+          window.setTimeout(() => this.next(), this.autoplaySpeed);
+    }
+  }
+
+  /**
+   * Run the loop to do all the necessary work for the carousel.
+   */
+  private onRaf(): void {
+    this.raf.read(() => {
+      if (!this.condition()) {
+        return;
+      }
+
+      if (this.isBeingInteractedWith()) {
+        this.resetAutoplayTimeout();
+      }
+
+      const activeSlide = this.getActiveSlide();
+      if (activeSlide !== this.lastActiveSlide) {
+        this.lastActiveSlide = activeSlide;
+        if (activeSlide) {
+          this.updateClasses(activeSlide);
+        }
+      }
+
+      if (this.transitionTarget !== null) {
+        const hasTransitionedToTarget =
+            this.transition.hasTransitionedTo(this.transitionTarget);
+        if (hasTransitionedToTarget) {
+          this.transitionTarget = null;
+          this.eventManager.dispatch(
+              CarouselEvent.AFTER_CHANGE, this.generateEventData());
```

Comment:
I would suggest having different data objects for beforeChange and for afterChange, since in the "after" event, I would think the `currentSlide` is what it has already transitioned to and `nextSlide` is what it will be when the carousel transitions in X seconds.

(In the "after" event maybe it should only include `currentSlide`.)

Related: is it worth adding the index to this object? For most carousels that I've worked on in the past, knowing the current index number is probably more important than having the slide element.

Reply:
Done.